provider/triton: New provider for Joyent Triton

jen20 opened the pull request:
This brings across the following resources for Triton from the joyent/triton-terraform repository, and converts them to the canonical Terraform style, introducing Terraform-style documentation and acceptance tests which run against the live API rather than the local APIs:

- triton_firewall_rule
- triton_machine
- triton_key

cc: @misterbisson, @brianhicks, @phinze, @stack72, @sigil66

stack72 commented on builtin/providers/triton/resource_firewall_rule.go:
```diff
@@ -0,0 +1,109 @@
+				Description: "Indicates if the rule is enabled",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
```
IMO this should mirror the Joyent default - this is what we tend to do with other providers AFAICT

jen20 replied:
Yes - my feeling is that we should stick to the API here. I'll remove the comment.

stack72 commented on builtin/providers/triton/resource_machine_test.go:
```diff
@@ -0,0 +1,121 @@
+	})
+}
+
+// func TestAccTritonMachine_update(t *testing.T) {
```
Any reason we are leaving this in?

stack72 commented:
This looks good - just need to fix the build and then its good to merge :)

jen20 commented:
I'm going to park this for now so that we can get 0.6.12 out the door - I'll come back around to it for 0.6.13. Seeing some weird issues with the SDC API and keys.

jen20 commented:
```
$ make testacc TEST=./builtin/providers/triton
==> Checking that code complies with gofmt requirements...
/Users/James/Code/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/triton -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccTritonFirewallRule_basic
--- PASS: TestAccTritonFirewallRule_basic (10.61s)
=== RUN   TestAccTritonFirewallRule_update
--- PASS: TestAccTritonFirewallRule_update (20.95s)
=== RUN   TestAccTritonFirewallRule_enable
--- PASS: TestAccTritonFirewallRule_enable (15.87s)
=== RUN   TestAccTritonKey_basic
--- PASS: TestAccTritonKey_basic (40.29s)
=== RUN   TestAccTritonMachine_basic
--- PASS: TestAccTritonMachine_basic (207.59s)
PASS
ok	    github.com/hashicorp/terraform/builtin/providers/triton	295.31s
```

phinze commented on builtin/providers/triton/provider.go:
```diff
@@ -0,0 +1,130 @@
+				Type:        schema.TypeString,
+				Required:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SDC_KEY_MATERIAL", ""),
+			},
```
Looks like this was never wired in?

phinze replied:
Following our pattern in other Terraform providers - we should probably switch to _just_ `key_material` and let `file()` pull in a path if necessary.

BrianHicks replied:
Agreed. I added this when first doing the provider and later changed it, this is a vestigal remnant from that. Looks like it can be removed with impunity. :)

phinze replied:
So based on my follow up comment we should switch to key_material and drop key_path. We're trying to consistently accept values and let paths always be pulled in the same way - via `file()`.

Example of this change in the `google` provider, where we deprecated `account_file` and moved to `credentials`:

https://github.com/hashicorp/terraform/blob/9202bc0e1cf534e2c1992cb0c1c8358b0c195ef0/builtin/providers/google/provider.go#L16-L29

In acceptance tests we usually end up supporting a test-specific env var to suck a file from disk like so:

https://github.com/hashicorp/terraform/blob/9202bc0e1cf534e2c1992cb0c1c8358b0c195ef0/builtin/providers/google/provider_test.go#L33-L39

The reasoning for this pattern is to properly support build environments (Atlas, Jenkins, other CI) where a file on disk is not always easy to come by, but env vars are supported.

Let me know if this makes sense! :+1:

BrianHicks replied:
Yes, that makes sense to me! Agreed about CI environments. :)

jen20 replied:
This is now the default - `key_path` has gone.

phinze commented on builtin/providers/triton/resource_firewall_rule_test.go:
```diff
@@ -0,0 +1,152 @@
+			resource.TestStep{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckTritonFirewallRuleExists("triton_firewall_rule.test"),
```
Definitely not a dealbreaker, but noting that we're missing a `testCheckTritonFirewallRuleAttributes` func here, so we're not verifying that the attributes are making it up to the API, only that they make it into the state.

phinze commented on builtin/providers/triton/resource_firewall_rule.go:
```diff
@@ -0,0 +1,103 @@
+		return err
+	}
+
+	d.SetId("")
```
Turns out these are unnecessary when `Delete` returns successfully. We only recently realized that we were doing this wrong in most places though, and it definitely doesn't hurt! :grinning:

phinze commented:
Just got enough elbow room to do a late review on this.

Everything looks great! Just that one thing about the key material I think is worth addressing before merge. (Though I'm willing to be overruled. :grinning:)

BrianHicks commented on builtin/providers/triton/resource_machine.go:
```diff
@@ -0,0 +1,438 @@
+	d.Set("image", machine.Image)
+	d.Set("primaryip", machine.PrimaryIP)
+	d.Set("networks", machine.Networks)
+	// d.Set("firewall_enabled", machine.FirewallEnabled) // but that field doesn't exist...
```
`machine.FirewallEnabled` is a field now, we added it so this'd work. This should be able to be uncommented.

jen20 commented:
@phinze I agree with you on key material. I think originally it was like that and I added back path to make the acceptance tests slightly easier to run locally and forgot to remove it again...

cmosetick commented:
Hi @jen20 Is this still scheduled to be included with 0.6.13?

phinze commented on builtin/providers/triton/resource_machine.go:
```diff
@@ -0,0 +1,436 @@
+				// TODO: this really should ForceNew but the Network IDs don't seem to
+				// be returned by the API, meaning if we track them here TF will replace
+				// the resource on every run.
+				// ForceNew:    true,
```
This comment doesn't sound like it's describing `ForceNew` but rather `Computed`?

phinze commented on builtin/providers/triton/resource_machine.go:
```diff
@@ -0,0 +1,436 @@
+	d.Set("package", machine.Package)
+	d.Set("image", machine.Image)
+	d.Set("primaryip", machine.PrimaryIP)
+	d.Set("networks", machine.Networks)
```
Probably worth checking the error return here since the value is not primitive, particularly given the comment above on the schema.

phinze commented on vendor/github.com/joyent/gocommon/COPYING:
```diff
@@ -0,0 +1,674 @@
+any implied license or other defenses to infringement that may
+otherwise be available to you under applicable patent law.
+
+  12. No Surrender of Others' Freedom.
```
I'm not so sure about this clause!




..just kidding. :stuck_out_tongue_closed_eyes:

phinze commented:
My last inline pair of comments are pretty minor and could be figured out post merge.

This LGTM. :ship: :rocket: :+1: :tada:

BrianHicks commented:
The comments about network will actually be obsoleted by some improvements we'll be making soon after merge.

> On Mar 18, 2016, at 6:38 PM, Paul Hinze <notifications@github.com> wrote:
> 
> My last inline pair of comments are pretty minor and could be figured out post merge.
> 
> This LGTM.    
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

jen20 commented:
Just done another run of the acceptance tests here and they look good to me:

```
make testacc TEST=./builtin/providers/triton
==> Checking that code complies with gofmt requirements...
/Users/James/Code/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/triton -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccTritonFirewallRule_basic
--- PASS: TestAccTritonFirewallRule_basic (8.02s)
=== RUN   TestAccTritonFirewallRule_update
--- PASS: TestAccTritonFirewallRule_update (13.82s)
=== RUN   TestAccTritonFirewallRule_enable
--- PASS: TestAccTritonFirewallRule_enable (13.22s)
=== RUN   TestAccTritonKey_basic
--- PASS: TestAccTritonKey_basic (18.34s)
=== RUN   TestAccTritonMachine_basic
--- PASS: TestAccTritonMachine_basic (211.27s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/triton	264.673s
```

I'll merge this for now and come back to the network-related issues when we get word from @misterbisson/@BrianHicks that there are upstream changes.

jen20 commented:
Just noticed that one of the dependencies, `joyent/gosign` is GPLv3 licensed so I've reverted this for now, and opened up and issue https://github.com/joyent/gosign/issues/4 to try to get this changed. There is some precedent here: `joyent/gosdc` itself was previously licensed as GPLv3 and changed to MPLv2 (https://github.com/joyent/gosdc/issues/11 indicated that GPLv3 is not used for new works).

misterbisson commented:
Good catch @jen20. I'll review and fix asap.

misterbisson commented:
PR opened in https://github.com/joyent/gosign/pull/5.

misterbisson commented:
I pinged a few people and heard no objections, so I merged https://github.com/joyent/gosign/pull/5 to relicense that repo as MPLv2.

lloydde commented:
@misterbisson you missed a spot ;-)

> Licensed under LGPL v3.
https://github.com/joyent/gosign/blob/master/gosign.go#L9

jen20 commented:
This was reverted, I'm going to open a new PR to get the correct bits with all the updated licenses referenced above merged in.
